### PR TITLE
user: fix removing the expiry time when it's 0

### LINKED ISSUE #47114: user: removal of the expiry time not working
##### SUMMARY
A negative value for the "expires" parameter of the user module should remove the expiration time of an account. This is not working.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
user

##### ANSIBLE VERSION

```paste below
ansible 2.7.0
  config file = /home_local/ansible/automation/ansible.cfg
  configured module search path = [u'/home_local/ansible/automation/library']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.5 (default, Jul 13 2018, 13:06:57) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```

##### CONFIGURATION

```paste below
ANSIBLE_NOCOWS(/home_local/ansible/automation/ansible.cfg) = False
DEFAULT_CALLBACK_WHITELIST(/home_local/ansible/automation/ansible.cfg) = [u'profile_tasks']
DEFAULT_FORKS(/home_local/ansible/automation/ansible.cfg) = 50
DEFAULT_GATHER_SUBSET(/home_local/ansible/automation/ansible.cfg) = !hardware
DEFAULT_HOST_LIST(/home_local/ansible/automation/ansible.cfg) = [u'/home_local/ansible/automation/inventory']
DEFAULT_LOG_PATH(/home_local/ansible/automation/ansible.cfg) = /home_local/ansible/automation/ansible.log
DEFAULT_MODULE_PATH(/home_local/ansible/automation/ansible.cfg) = [u'/home_local/ansible/automation/library']
DEFAULT_ROLES_PATH(/home_local/ansible/automation/ansible.cfg) = [u'/home_local/ansible/automation/roles']
DEFAULT_TIMEOUT(/home_local/ansible/automation/ansible.cfg) = 60
DEFAULT_UNDEFINED_VAR_BEHAVIOR(/home_local/ansible/automation/ansible.cfg) = True
DEFAULT_VAULT_PASSWORD_FILE(/home_local/ansible/automation/ansible.cfg) = /home_local/ansible/vault/vault.key
HOST_KEY_CHECKING(/home_local/ansible/automation/ansible.cfg) = True
INVENTORY_IGNORE_EXTS(/home_local/ansible/automation/ansible.cfg) = [u'~', u'.orig', u'.bak', u'.ini', u'.cfg', u'.retry', u
RETRY_FILES_ENABLED(/home_local/ansible/automation/ansible.cfg) = False
```

##### OS / ENVIRONMENT

Seen on CentOS 7.5:

```
$ cat /etc/redhat-release 
CentOS Linux release 7.5.1804 (Core) 
```

##### STEPS TO REPRODUCE
Starting from a user which account is about to expire, we run following playbook to remove the expiration time:

test.yml:

```yaml
- hosts: localhost
  become: true
  become_user: root

  tasks:
    - user:
        name: test_expire
        expires: -1
```

##### EXPECTED RESULTS

Starting situation:

```
$sudo chage -l test_expire
Last password change                                    : Oct 16, 2018
Password expires                                        : Jan 14, 2019
Password inactive                                       : never
Account expires                                         : Jan 01, 1970
Minimum number of days between password change          : 0
Maximum number of days between password change          : 90
Number of days of warning before password expires       : 14
```

Expected results after running `ansible-playbook test.yml`:

```
$ sudo chage -l test_expire
Last password change                                    : Oct 16, 2018
Password expires                                        : Jan 14, 2019
Password inactive                                       : never
Account expires                                         : never
Minimum number of days between password change          : 0
Maximum number of days between password change          : 90
Number of days of warning before password expires       : 14
```

Note the "Account expires" row.

##### ACTUAL RESULTS

Actual results after running `ansible-playbook test.yml`:

```
$ sudo chage -l test_expire
Last password change                                    : Oct 16, 2018
Password expires                                        : Jan 14, 2019
Password inactive                                       : never
Account expires                                         : Jan 01, 1970
Minimum number of days between password change          : 0
Maximum number of days between password change          : 90
Number of days of warning before password expires       : 14
```

### PR DESCRIPTION
**Pull request by mkrizek**
##### SUMMARY

Fixes https://github.com/ansible/ansible/issues/47114
##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

user

##### ANSIBLE VERSION

```paste below
2.7
devel
```

### CODE REVIEW

File: test/integration/targets/user/tasks/main.yml [hunk 1]
```diff
@@ -429,6 +429,52 @@
...
+- name: Remove ansibulluser
+  user:
+    name: ansibulluser
+    state: absent
+
+- name: Create user account with epoch 0 expiration
+  user:
+    name: ansibulluser
+    state: present
+    expires: 0
```
Comment (samdoran) [inline test/integration/targets/user/tasks/main.yml:479]: Run this again to test no changes are reported. Currently, it always reports a change every time.

File: test/integration/targets/user/tasks/main.yml [hunk 2]
```diff
@@ -429,6 +429,52 @@
...
+
+- name: Change the user account to remove the expiry time
+  user:
+    name: ansibulluser
+    expires: -1
```
Comment (samdoran) [inline test/integration/targets/user/tasks/main.yml:492]: Run this again to test no changes are reported. Currently, it always reports a change every time.

**Review by samdoran — CHANGES_REQUESTED** [review]
This is only part of the fix. Zero is tricky. Lines 709 and 1139 need to change frmo `<= 0` to `< 0`. Otherwise, a change is made every time.

```python
if current_expires < 0 or current_expire_date[:3] != self.expires[:3]:
    cmd.append('-e')
    cmd.append(time.strftime(self.DATE_FORMAT, self.expires))
```

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/88671/32/tests

**Comment by samdoran** [comment]
Oh, that's good. Looks like we have tests that caught that. :)

File: test/integration/targets/user/tasks/main.yml (hunk 1 above)
Comment (mkrizek) [reply test/integration/targets/user/tasks/main.yml:479]: Seems like it happens only on FreeBSD though? :/

Comment (samdoran) [reply test/integration/targets/user/tasks/main.yml:479]: I was able to get it to report changes every time testing on a Linux VM.

Comment (mkrizek) [reply test/integration/targets/user/tasks/main.yml:479]: OK, I was able to reproduce it on Linux as well. Changed the fix as per your suggestion. I left the FreeBSD portion of the code intact as the issue this is fixing is reported against Linux.

**Review by samdoran — APPROVED** [review]

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/92197/33/tests

**Comment by mkrizek** [comment]
@samdoran looks like with the integration test you added revealed another bug that is not a result of this PR as it fails even without this change. Should we deal with that separately?

**Comment by samdoran** [comment]
@mkrizek I don't think it's a bug in the module but a bug in the test.

I dug into a bit and on FreeBSD, setting expires to `0` disables expiration. That means when running the task with `expires: 0` and `expires: -1` are the same and result in the same end state of the system. 

I changed the tests to check  for different results based on BSD or Linux. I also changed the expiration logic slightly in the OpenBSD class since `-1` and `0` are the same thing on the underlying system and should result in the same command.

**Comment by mkrizek** [comment]
Thanks @samdoran!

**Comment by samdoran** [comment]
Thank you @mkrizek for the initial fix. 🍻

**Pull request merged**